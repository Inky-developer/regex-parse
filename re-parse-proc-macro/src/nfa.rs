use crate::arena::{Arena, ArenaIndex};
use crate::regex::{Regex, RegexArena, RegexNode, RegexNodeIndex, RegexPattern, RegexVariable};
use crate::util::FloodFill;
use crate::Set;

pub type NfaArena = Arena<NfaNode>;
pub type NfaIndex = ArenaIndex<NfaNode>;

#[derive(Debug)]
pub struct Nfa {
    pub root: NfaIndex,
    pub nodes: NfaArena,
}

impl From<Regex> for Nfa {
    fn from(value: Regex) -> Self {
        let Regex { arena, root } = value;
        let mut nodes = NfaArena::default();
        let root_node = nodes.add(NfaNode::EPSILON);
        let target_node = convert_regex_node(&mut nodes, &arena, root, root_node);
        nodes[target_node].is_accepting = true;

        check_variables(&nodes);

        Nfa {
            nodes,
            root: root_node,
        }
    }
}

fn check_variables(nodes: &NfaArena) {
    let mut visited_variables = Set::default();
    for node in nodes.iter() {
        if let NfaNodeKind::Variable(RegexVariable { name, .. }) = &nodes[node].kind {
            if visited_variables.contains(name) {
                panic!(
                    "(TODO MAKE ERROR) Variable \"{}\" is already declared",
                    name
                );
            }
            visited_variables.insert(name.clone());
        }
    }
}

#[derive(Debug)]
pub struct NfaNode {
    pub edges: Vec<NfaIndex>,
    pub edge_kind: NfaEdge,
    pub kind: NfaNodeKind,
    pub is_accepting: bool,
}

impl NfaNode {
    pub const EPSILON: Self = Self {
        edges: Vec::new(),
        edge_kind: NfaEdge::Epsilon,
        kind: NfaNodeKind::Simple,
        is_accepting: false,
    };
}

#[derive(Debug)]
pub enum NfaNodeKind {
    Simple,
    Variable(RegexVariable),
}

#[derive(Debug)]
pub enum NfaEdge {
    Epsilon,
    Pattern(RegexPattern),
}

impl NfaEdge {
    pub fn is_epsilon(&self) -> bool {
        matches!(self, NfaEdge::Epsilon)
    }
}

fn convert_regex_node(
    arena: &mut NfaArena,
    regex_arena: &RegexArena,
    node: RegexNodeIndex,
    predecessor: NfaIndex,
) -> NfaIndex {
    let node = &regex_arena[node];
    match node {
        RegexNode::And(nodes) => {
            let mut last_node = predecessor;
            for node in nodes {
                let new_node = convert_regex_node(arena, regex_arena, *node, last_node);
                last_node = new_node;
            }
            last_node
        }
        RegexNode::Or(nodes) => {
            let target_node = arena.add(NfaNode::EPSILON);
            for node in nodes {
                let new_node = convert_regex_node(arena, regex_arena, *node, predecessor);
                arena.connect(new_node, target_node);
            }
            target_node
        }
        RegexNode::Literal(pattern) => arena.add_after(
            predecessor,
            NfaNode {
                edges: Vec::new(),
                edge_kind: NfaEdge::Pattern(*pattern),
                kind: NfaNodeKind::Simple,
                is_accepting: false,
            },
        ),
        RegexNode::Variable(var) => {
            let node = arena.add_after(
                predecessor,
                NfaNode {
                    edges: Vec::new(),
                    edge_kind: NfaEdge::Pattern(RegexPattern::AnyCharLazy),
                    kind: NfaNodeKind::Variable(var.clone()),
                    is_accepting: false,
                },
            );
            arena.connect(node, node);
            node
        }
        RegexNode::ZeroOrOne(child) => {
            let target_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, target_node);
            let new_node = convert_regex_node(arena, regex_arena, *child, predecessor);
            arena.connect(new_node, target_node);
            target_node
        }
        RegexNode::Many(child) => {
            let iteration_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, iteration_node);
            let target_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, target_node);
            let new_node = convert_regex_node(arena, regex_arena, *child, iteration_node);
            arena.connect(new_node, iteration_node);
            arena.connect(new_node, target_node);
            target_node
        }
        RegexNode::OneOrMore(child) => {
            let iteration_node = arena.add(NfaNode::EPSILON);
            arena.connect(predecessor, iteration_node);
            let target_node = arena.add(NfaNode::EPSILON);
            let new_node = convert_regex_node(arena, regex_arena, *child, iteration_node);
            arena.connect(new_node, iteration_node);
            arena.connect(new_node, target_node);
            target_node
        }
    }
}

impl NfaArena {
    fn connect(&mut self, source: NfaIndex, target: NfaIndex) {
        self[source].edges.push(target);
    }

    fn add_after(&mut self, source: NfaIndex, new_node: NfaNode) -> NfaIndex {
        let idx = self.add(new_node);
        self.connect(source, idx);
        idx
    }
}

impl FloodFill for Nfa {
    type Item = NfaIndex;

    fn get_neighbors(&self, item: &Self::Item) -> impl Iterator<Item = Self::Item> {
        self.nodes[*item].edges.iter().copied()
    }
}

#[cfg(test)]
mod tests {
    use crate::nfa::Nfa;
    use crate::parser::ParseError;
    use crate::regex::Regex;

    fn parse(source: &str) -> Result<Nfa, ParseError> {
        let regex = Regex::from_str(source)?;
        Ok(Nfa::from(regex))
    }

    #[test]
    fn test_regex_to_nfa() {
        insta::assert_debug_snapshot!(parse("A"));
        insta::assert_debug_snapshot!(parse("A|B|C"));
        insta::assert_debug_snapshot!(parse("A?b*c"));
        insta::assert_debug_snapshot!(parse(".{var}."));
        insta::assert_debug_snapshot!(parse(".+;"));
    }
}
