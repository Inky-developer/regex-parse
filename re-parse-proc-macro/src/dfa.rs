use crate::arena::{Arena, ArenaIndex};
use crate::nfa::{Nfa, NfaEdge, NfaIndex, NfaNodeKind};
use crate::regex::{RegexPattern, RegexVariable};
use crate::util::FloodFill;
use crate::{Map, Set};

pub type DfaArena = Arena<DfaNode>;
pub type DfaIndex = ArenaIndex<DfaNode>;

#[derive(Debug)]
pub struct Dfa {
    pub root: DfaIndex,
    pub nodes: DfaArena,
}

impl Dfa {
    pub fn iter(&self) -> impl Iterator<Item = DfaIndex> + use<'_> {
        <Self as FloodFill>::iter(self, self.root)
    }
}

impl From<Nfa> for Dfa {
    fn from(nfa: Nfa) -> Self {
        let mut builder = DfaBuilder::default();
        let hull = calculate_epsilon_hull(&nfa);
        let root_group = hull[&nfa.root].clone();
        builder.pending_nodes.extend(hull.into_values());

        while let Some(group) = builder.pending_nodes.iter().next() {
            let group = group.clone();
            builder.pending_nodes.remove(&group);

            builder.compute_group(&nfa, group);
        }

        let root = builder.nfa_to_dfa[&root_group];
        Dfa {
            root,
            nodes: builder.nodes,
        }
    }
}

impl FloodFill for Dfa {
    type Item = DfaIndex;

    fn get_neighbors(&self, item: &Self::Item) -> impl Iterator<Item = Self::Item> {
        let edges = &self.nodes[*item].edges;
        edges
            .default
            .iter()
            .copied()
            .chain(edges.edges.values().copied())
    }
}

#[derive(Debug, Default)]
pub struct DfaBuilder {
    nodes: DfaArena,
    nfa_to_dfa: Map<Vec<NfaIndex>, DfaIndex>,
    pending_nodes: Set<Vec<NfaIndex>>,
}

impl DfaBuilder {
    fn insert(&mut self, key: Vec<NfaIndex>, node: DfaNode) -> DfaIndex {
        if let Some(idx) = self.nfa_to_dfa.get(&key) {
            self.nodes[*idx] = node;
            return *idx;
        }

        let idx = self.nodes.add(node);
        self.nfa_to_dfa.insert(key, idx);
        idx
    }

    fn entry(&mut self, key: Vec<NfaIndex>) -> DfaIndex {
        if let Some(idx) = self.nfa_to_dfa.get(&key) {
            return *idx;
        }

        let node = DfaNode::default();
        self.pending_nodes.insert(key.clone());
        self.insert(key, node)
    }

    fn compute_group(&mut self, nfa: &Nfa, group: Vec<NfaIndex>) {
        let edges = DfaEdges::from_nfa_group(self, nfa, &group);
        let is_accepting = group
            .iter()
            .copied()
            .any(|nfa_idx| nfa.nodes[nfa_idx].is_accepting);
        let variable = self.compute_group_variable(nfa, &group);

        self.insert(
            group,
            DfaNode {
                is_accepting,
                variable,
                edges,
            },
        );
    }

    fn compute_group_variable(&self, nfa: &Nfa, group: &[NfaIndex]) -> Option<RegexVariable> {
        let mut variable = None;

        for nfa_idx in group.iter().copied() {
            let NfaNodeKind::Variable(var) = &nfa.nodes[nfa_idx].kind else {
                continue;
            };

            match variable {
                None => variable = Some(var.clone()),
                Some(RegexVariable {
                    name: other_var, ..
                }) => panic!(
                    "(TODO THROW ERROR) ambiguous variables: {other_var} collides with {}",
                    &var.name
                ),
            }
        }

        variable
    }
}

fn get_non_epsilon_edges(nfa: &Nfa, group: &[NfaIndex]) -> Vec<(RegexPattern, NfaIndex)> {
    let mut edges: Vec<(RegexPattern, NfaIndex)> = Vec::new();
    for node_idx in group {
        let node = &nfa.nodes[*node_idx];
        for edge_idx in &node.edges {
            let edge = &nfa.nodes[*edge_idx];
            if let NfaEdge::Pattern(pattern) = &edge.edge_kind {
                edges.push((*pattern, *edge_idx))
            }
        }
    }
    edges
}

fn calculate_epsilon_hull(nfa: &Nfa) -> Map<NfaIndex, Vec<NfaIndex>> {
    let mut hull = Map::default();

    for idx in nfa.iter() {
        hull.insert(idx, get_connected_nodes(nfa, idx));
    }

    hull
}

fn expand_group(nfa: &Nfa, group: &[NfaIndex]) -> Vec<NfaIndex> {
    let mut nodes = Set::default();
    for idx in group.iter().copied() {
        nodes.extend(get_connected_nodes(nfa, idx));
    }

    let mut result = nodes.into_iter().collect::<Vec<_>>();
    result.sort();
    result
}

fn get_connected_nodes(nfa: &Nfa, idx: NfaIndex) -> Vec<NfaIndex> {
    let mut nodes: Set<NfaIndex> = Set::default();
    let mut pending_nodes: Set<NfaIndex> = Set::default();

    pending_nodes.insert(idx);
    while let Some(node) = pending_nodes.iter().copied().next() {
        pending_nodes.remove(&node);
        nodes.insert(node);

        pending_nodes.extend(
            nfa.nodes[node]
                .edges
                .iter()
                .copied()
                .filter(|edge| nfa.nodes[*edge].edge_kind.is_epsilon()),
        )
    }

    let mut result: Vec<NfaIndex> = nodes.into_iter().collect();
    result.sort();
    result
}

#[derive(Debug, Default)]
pub struct DfaNode {
    pub is_accepting: bool,
    pub variable: Option<RegexVariable>,
    pub edges: DfaEdges,
}

#[derive(Debug, Default)]
pub struct DfaEdges {
    pub default: Option<DfaIndex>,
    pub edges: Map<char, DfaIndex>,
}

impl DfaEdges {
    fn from_nfa_group(dfa: &mut DfaBuilder, nfa: &Nfa, group: &[NfaIndex]) -> Self {
        let edges = get_non_epsilon_edges(nfa, group);

        let mut default_edges: Vec<NfaIndex> = Vec::new();
        let mut lazy_default_edges: Vec<NfaIndex> = Vec::new();

        let mut edge_map: Map<char, Vec<NfaIndex>> = Map::default();
        for (edge_pattern, target_idx) in edges.iter().copied() {
            match edge_pattern {
                RegexPattern::Char(char) => edge_map.entry(char).or_default().push(target_idx),
                RegexPattern::Range(start, end) => {
                    for char in start..=end {
                        edge_map.entry(char).or_default().push(target_idx);
                    }
                }
                RegexPattern::AnyChar => default_edges.push(target_idx),
                RegexPattern::AnyCharLazy => lazy_default_edges.push(target_idx),
            }
        }

        // Since a default edge can be any char, it also has to be added to each value in the edge map now.
        for targets in edge_map.values_mut() {
            targets.extend(default_edges.iter().copied());
            targets.sort_unstable();
            targets.dedup();
        }

        // If there is a default_edge, it will overwrite the lazy-default edge completely.
        if default_edges.is_empty() {
            default_edges = lazy_default_edges;
        }
        default_edges.sort_unstable();
        default_edges.dedup();

        let default_edge_idx = if default_edges.is_empty() {
            None
        } else {
            Some(dfa.entry(expand_group(nfa, &default_edges)))
        };
        let edge_indices = edge_map
            .into_iter()
            .map(|(key, value)| (key, dfa.entry(expand_group(nfa, &value))))
            .collect();
        DfaEdges {
            default: default_edge_idx,
            edges: edge_indices,
        }
    }
}

#[cfg(test)]
mod tests {
    use crate::dfa::Dfa;
    use crate::nfa::Nfa;
    use crate::parser::ParseError;
    use crate::regex::Regex;

    fn parse(input: &str) -> Result<Dfa, ParseError> {
        let regex = Regex::from_str(input)?;
        let nfa = Nfa::from(regex);
        Ok(Dfa::from(nfa))
    }

    #[test]
    fn test_nfa_to_dfa() {
        insta::assert_debug_snapshot!(parse("A"));
        insta::assert_debug_snapshot!(parse("AB"));
        insta::assert_debug_snapshot!(parse("A?B"));
        insta::assert_debug_snapshot!(parse("A?A"));
        insta::assert_debug_snapshot!(parse("A?b*c"));
        insta::assert_debug_snapshot!(parse("{foo}"));
        insta::assert_debug_snapshot!(parse("A{foo}B+{bar}"));
        insta::assert_debug_snapshot!(parse("[a-e]"));
        insta::assert_debug_snapshot!(parse(".{var}."));
    }

    #[test]
    #[should_panic(expected = "(TODO THROW ERROR) ambiguous variables: foo collides with bar")]
    fn test_nfa_to_dfa_ambiguous_variable() {
        parse("A{foo}B?{bar}").unwrap();
    }
}
