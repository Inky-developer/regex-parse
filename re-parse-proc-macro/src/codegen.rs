use crate::dfa::{Dfa, DfaIndex};
use crate::regex::VariableKind;
use crate::{Map, Set};
use proc_macro2::{Ident, Span, TokenStream};
use quote::quote;
use syn::Expr;

pub struct Codegen {
    pub dfa: Dfa,
    pub expression: Expr,
}

impl Codegen {
    pub fn generate(self) -> TokenStream {
        let variables = self.collect_variables();
        let variable_idents = variables
            .iter()
            .enumerate()
            .map(|(index, _)| Ident::new(&format!("__var_{index}"), Span::mixed_site()))
            .collect::<Vec<_>>();
        let variable_map = variables
            .iter()
            .zip(variable_idents.iter())
            .map(|(var, ident)| {
                (
                    var.ident.to_string(),
                    Variable {
                        ident: ident.clone(),
                        kind: var.kind,
                    },
                )
            })
            .collect::<Map<_, _>>();

        let variable_setups = variable_map
            .values()
            .map(|var| self.quote_variable_setup(var));
        let variable_finalizers = variable_map
            .iter()
            .map(|(k, v)| self.quote_variable_finalizer(v, k));

        let states = self.collect_states();
        let internal_states = states.values();
        let initial_state = &states[&self.dfa.root];

        let state_branches = self.collect_state_branches(&states, &variable_map);
        let state_terminations = self.collect_state_terminations(&states, &variable_map);

        let expr = &self.expression;

        quote! {
            {
                #(#variable_setups)*

                enum __State {
                    #(#internal_states),*
                }

                let __initial_input = #expr;
                let mut __input = __initial_input.char_indices();
                let mut __variable_start = 0_usize;

                let mut __state = __State::#initial_state;
                loop {
                    let Some((__byte_index, __next_char)) = __input.next() else {
                        match __state {
                            #(#state_terminations),*
                        }
                    };
                    match __state {
                        #(#state_branches),*
                    }
                }

                #(#variable_finalizers)*
            }
        }
    }

    fn quote_variable_finalizer(&self, var: &Variable, name: &str) -> TokenStream {
        let ident = &var.ident;
        let original_ident = Ident::new(name, Span::call_site());
        match var.kind {
            VariableKind::Singular => {
                quote! { #original_ident = __initial_input[#ident].parse().unwrap();}
            }
            VariableKind::Multiple => {
                quote! { #original_ident = #ident.into_iter().map(|span| __initial_input[span].parse().unwrap()).collect(); }
            }
        }
    }

    fn quote_variable_setup(&self, var: &Variable) -> TokenStream {
        let ident = &var.ident;
        match var.kind {
            VariableKind::Singular => quote! { let mut #ident = 0_usize..0; },
            VariableKind::Multiple => quote! { let mut #ident = ::std::vec::Vec::new(); },
        }
    }

    fn collect_state_terminations(
        &self,
        states: &Map<DfaIndex, Ident>,
        variables: &Map<String, Variable>,
    ) -> Vec<TokenStream> {
        states
            .iter()
            .map(|(dfa_idx, internal_name)| {
                self.collect_state_termination(*dfa_idx, internal_name, variables)
            })
            .collect()
    }

    fn collect_state_termination(
        &self,
        dfa_idx: DfaIndex,
        internal_name: &Ident,
        variables: &Map<String, Variable>,
    ) -> TokenStream {
        let state = &self.dfa.nodes[dfa_idx];

        let panic_message = format!("Unexpected end of input ({internal_name})");

        let termination = match (state.is_accepting, &state.variable) {
            (true, Some(var)) => {
                let internal_var = &variables[&var.name];
                let update =
                    self.quote_update_variable(internal_var, quote! {__initial_input.len()});
                quote! {
                    {
                        #update;
                        break;
                    }
                }
            }
            (true, None) => quote! { break },
            (false, _) => quote! {panic!(#panic_message)},
        };

        quote! {
            __State::#internal_name => #termination
        }
    }

    fn quote_update_variable(&self, variable: &Variable, variable_end: TokenStream) -> TokenStream {
        let ident = &variable.ident;
        match variable.kind {
            VariableKind::Singular => quote! { #ident = __variable_start..#variable_end; },
            VariableKind::Multiple => {
                quote! { #ident.push(__variable_start..#variable_end); }
            }
        }
    }

    fn collect_state_branches(
        &self,
        states: &Map<DfaIndex, Ident>,
        variables: &Map<String, Variable>,
    ) -> Vec<TokenStream> {
        // Let's sort the states first to make it easier to read the macro expansion
        let mut sorted_states = states.iter().collect::<Vec<_>>();
        sorted_states.sort_unstable_by_key(|(_, name)| *name);

        sorted_states
            .iter()
            .map(|(dfa_idx, internal_name)| {
                self.collect_state_branch(**dfa_idx, internal_name, states, variables)
            })
            .collect()
    }

    fn collect_state_branch(
        &self,
        dfa_idx: DfaIndex,
        internal_name: &Ident,
        states: &Map<DfaIndex, Ident>,
        variables: &Map<String, Variable>,
    ) -> TokenStream {
        let state = &self.dfa.nodes[dfa_idx];

        let default_edge = match state.edges.default {
            Some(target) => (
                None,
                StateTransition::Valid {
                    target: states[&target].clone(),
                    variable_update: self.make_variable_update(dfa_idx, target, variables),
                },
            ),
            None => {
                let expected = if state.edges.edges.is_empty() {
                    vec!["End of input".to_string()]
                } else {
                    state.edges.edges.keys().copied().map(Into::into).collect()
                };
                (None, StateTransition::Invalid { expected })
            }
        };
        let initial_patterns = state
            .edges
            .edges
            .iter()
            .map(|(char, idx)| {
                (
                    Some(*char),
                    StateTransition::Valid {
                        target: states[idx].clone(),
                        variable_update: self.make_variable_update(dfa_idx, *idx, variables),
                    },
                )
            })
            .chain(std::iter::once(default_edge));

        let simplified_patterns = self.simplify_match(initial_patterns);

        quote! {
            __State::#internal_name => {
                match __next_char {
                    #(#simplified_patterns)*
                }
            }
        }
    }

    fn simplify_match(
        &self,
        patterns_and_transitions: impl Iterator<Item = (Option<char>, StateTransition)>,
    ) -> Vec<TokenStream> {
        let mut simplified: Map<StateTransition, Vec<Option<char>>> = Map::default();

        for (pattern, transition) in patterns_and_transitions {
            simplified
                .entry(transition.clone())
                .or_default()
                .push(pattern);
        }

        // Sort the patterns and transitions, so that the default pattern is always at the end
        let mut simplified: Vec<_> = simplified.into_iter().collect();
        simplified.sort_unstable_by_key(|(_, patterns)| patterns.iter().any(|it| it.is_none()));

        simplified
            .into_iter()
            .map(|(transition, patterns)| {
                let transition = transition.quote();
                if patterns.iter().any(|it| it.is_none()) {
                    quote! {_ => {
                        #transition
                    }}
                } else {
                    let chars = patterns.iter().map(|it| it.unwrap());
                    quote! {#(#chars)|* => #transition,}
                }
            })
            .collect()
    }

    fn make_variable_update(
        &self,
        current_idx: DfaIndex,
        target_idx: DfaIndex,
        variables: &Map<String, Variable>,
    ) -> VariableUpdate {
        let current_state = &self.dfa.nodes[current_idx];
        let target_state = &self.dfa.nodes[target_idx];

        match (&current_state.variable, &target_state.variable) {
            (None, Some(_)) => VariableUpdate::StartVariable,
            (Some(var), None) => VariableUpdate::EndVariable(variables[&var.name].clone()),
            _ => VariableUpdate::NoVariable,
        }
    }

    fn collect_variables(&self) -> Vec<Variable> {
        let mut variables = Set::default();
        for node_idx in self.dfa.iter() {
            let node = &self.dfa.nodes[node_idx];
            if let Some(variable) = &node.variable {
                let ident = Ident::new(&variable.name, Span::call_site());
                variables.insert(Variable {
                    ident,
                    kind: variable.kind,
                });
            }
        }

        variables.into_iter().collect()
    }

    fn collect_states(&self) -> Map<DfaIndex, Ident> {
        self.dfa
            .iter()
            .enumerate()
            .map(|(index, dfa_idx)| {
                (
                    dfa_idx,
                    Ident::new(&format!("State_{index}"), Span::mixed_site()),
                )
            })
            .collect()
    }
}

#[derive(Debug, Eq, PartialEq, Hash, Clone)]
struct Variable {
    kind: VariableKind,
    ident: Ident,
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
enum StateTransition {
    Invalid {
        expected: Vec<String>,
    },
    Valid {
        target: Ident,
        variable_update: VariableUpdate,
    },
}

impl StateTransition {
    fn quote(&self) -> TokenStream {
        match self {
            StateTransition::Invalid { expected } => {
                let message = match expected.as_slice() {
                    [single] => {
                        format!("Unexpected character {{__next_char}}. Expected '{single}'")
                    }
                    _ => format!(
                        "Unexpected character: {{__next_char}}. Expected one of: '{}'",
                        expected.join(", ")
                    ),
                };
                quote! {panic!(#message)}
            }
            StateTransition::Valid {
                target,
                variable_update,
            } => {
                let variable_update = variable_update.quote();
                quote! {{
                    #variable_update
                    __state = __State::#target;
                }}
            }
        }
    }
}

#[derive(Debug, Clone, Eq, PartialEq, Hash)]
enum VariableUpdate {
    NoVariable,
    StartVariable,
    EndVariable(Variable),
}

impl VariableUpdate {
    fn quote(&self) -> TokenStream {
        match self {
            VariableUpdate::NoVariable => quote! {},
            VariableUpdate::StartVariable => quote! {__variable_start = __byte_index;},
            VariableUpdate::EndVariable(Variable {
                kind: VariableKind::Singular,
                ident,
            }) => quote! {#ident = __variable_start..__byte_index;},
            VariableUpdate::EndVariable(Variable {
                kind: VariableKind::Multiple,
                ident,
            }) => quote! {#ident.push(__variable_start..__byte_index);},
        }
    }
}
