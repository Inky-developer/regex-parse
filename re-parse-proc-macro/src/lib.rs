mod arena;
mod codegen;
mod dfa;
mod nfa;
mod parser;
mod regex;
mod tokenizer;
mod util;

use crate::codegen::Codegen;
use crate::dfa::Dfa;
use crate::nfa::Nfa;
use crate::regex::Regex;
use proc_macro::TokenStream;
use syn::parse::{Parse, ParseStream};
use syn::{parse_macro_input, Expr, LitStr};

// Use non-std map and set implementations to make snapshot testing possible.
// std map and set implementations are not deterministic, which is required for that.
pub(crate) type Map<K, V> = fxhash::FxHashMap<K, V>;
pub(crate) type Set<K> = fxhash::FxHashSet<K>;

struct ReParseInput {
    regex: LitStr,
    expression: Expr,
}

impl Parse for ReParseInput {
    fn parse(input: ParseStream) -> syn::Result<Self> {
        let regex = input.parse()?;
        input.parse::<syn::Token![,]>()?;
        let expression = input.parse()?;
        Ok(Self { regex, expression })
    }
}

/// The main macro of this crate, which parses strings using regular expressions and can extract variables.
///
/// # Usage
/// `re_parse!(pattern: StrLiteral, value: &str);`
///
/// Any variables contained in `pattern` will be set after the macro has run.
/// For now, the macro will panic if the input cannot be parsed (TODO: Return error)
///
/// The pattern is a regular expression which can contain variable captures.
///
/// ## Variable Captures
/// - `{var_name}`: Captures a single variable of at least one character
/// - `{var_name*}`: Captures multiple (or zero) variables
///
/// ## Character Classes
/// `re_parse!` currently supports these character classes:
/// - `\s`: Any Whitespace (equivalent to `[\n\t\r ]`)
/// - `\d`: Any Digit (equivalent to `[0-9]`)
/// - `\w`: Any Word (equivalent to `[a-zA-Z0-0_]`)
///
/// # Example
///
/// ```rust
/// # use re_parse_proc_macro::re_parse;
/// let name: String;
/// let score: f32;
/// re_parse!("The score of {name} is {score}", "The score of user is 55.8");
/// assert_eq!(name, "user");
/// assert_eq!(score, 55.8);
/// ```
///
/// ## Multiple variables
/// ```rust
/// # use re_parse_proc_macro::re_parse;
/// let temperatures: Vec<f32>;
/// re_parse!(r"Temperatures: \[({temperatures*}\s*,?\s*)*\]", "Temperatures: [10.0, 9.0, 8.5, 8.0]");
/// assert_eq!(temperatures, vec![10.0, 9.0, 8.5, 8.0]);
/// ```
///
/// # Efficiency
/// The macro compiles the pattern into a state-machine which executes in linear time, so it should be very efficient.
#[proc_macro]
pub fn re_parse(input: TokenStream) -> TokenStream {
    let ReParseInput { regex, expression } = parse_macro_input!(input as ReParseInput);

    let result = re_parse_impl(regex, expression);
    result.into()
}

fn re_parse_impl(regex: LitStr, expression: Expr) -> proc_macro2::TokenStream {
    let regex = Regex::from_str(&regex.value()).unwrap();
    let nfa = Nfa::from(regex);
    let dfa = Dfa::from(nfa);
    let codegen = Codegen { dfa, expression };
    codegen.generate()
}

#[cfg(test)]
mod tests {
    use super::{re_parse_impl, ReParseInput};
    use quote::quote;

    fn test_re_parse(input: proc_macro2::TokenStream) -> String {
        let ReParseInput { regex, expression } = syn::parse2::<ReParseInput>(input).unwrap();
        let stream = re_parse_impl(regex, expression);
        let file_content = format!("fn main() {{ {stream} }}");
        let file = syn::parse_file(&file_content).unwrap();
        prettyplease::unparse(&file)
    }

    macro_rules! dbg_re_parse {
        ($($input:tt)*) => {test_re_parse(quote! {$($input)*})};
    }

    #[test]
    fn test_macro_expansion() {
        insta::assert_snapshot!(dbg_re_parse!("A", "A"));
        insta::assert_snapshot!(dbg_re_parse!("A+", "A"));
        insta::assert_snapshot!(dbg_re_parse!("({var*},)*", "1,2,3,4,"));
        insta::assert_snapshot!(dbg_re_parse!("([abc]\\s*)*", "A"));
        insta::assert_snapshot!(dbg_re_parse!("A.*B.*;", "AAABBB;"));
    }
}
