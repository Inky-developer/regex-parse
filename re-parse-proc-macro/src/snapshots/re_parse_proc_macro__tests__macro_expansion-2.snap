---
source: re-parse-proc-macro/src/lib.rs
expression: "dbg_re_parse!(\"A+\", \"A\")"
snapshot_kind: text
---
fn main() {
    {
        enum __State {
            State_1,
            State_0,
        }
        let __initial_input = "A";
        let mut __input = __initial_input.char_indices();
        let mut __variable_start = 0_usize;
        let mut __state = __State::State_0;
        loop {
            let Some((__byte_index, __next_char)) = __input.next() else {
                match __state {
                    __State::State_1 => break,
                    __State::State_0 => panic!("Unexpected end of input (State_0)"),
                }
            };
            match __state {
                __State::State_0 => {
                    match __next_char {
                        'A' => {
                            __state = __State::State_1;
                        }
                        char => panic!("Invalid character: {char}"),
                    }
                }
                __State::State_1 => {
                    match __next_char {
                        'A' => {
                            __state = __State::State_1;
                        }
                        char => panic!("Invalid character: {char}"),
                    }
                }
            }
        }
    }
}
